src/gt4py/next/iterator/transforms/trace_shifts.py

```diff
@@ -117,16 +67,34 @@ def apply(arg):
     return apply
 
 
+@dataclass(frozen=True)
+class AppliedLift:
+    stencil: Callable
+    its: List[Union[InputTracer, "AppliedLift"]]
+
+    def __post_init__(self):
+        for it in self.its:
+            assert isinstance(it, (InputTracer, AppliedLift))
```

Comment:
This looks very conservative to me. What about doing this check before construction if it's really needed? Or no check, and instead assume type_inference works as a pre-condition of this pass.

Reply:
Doing it in the lift looks nice, I still kept the assert both for safety and clarity.